components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/util/ClientCredentialAuthConfig.java

```diff
@@ -0,0 +1,320 @@
+    private AtomicInteger tokenRequestAttemptCountForTimeOut;
+    private APIAccessTokenCache apiAccessTokenCache;
+    private String consumerKey;
+    private String consumerSecret;
+    private String scopes;
+    private String tokenEndpoint;
+    private AuthenticationContext authenticationContext;
+    private AsyncReturn asyncReturn;
+
+    public void setAuthenticationContext(AuthenticationContext authenticationContext) {
+        this.authenticationContext = authenticationContext;
+    }
+
+    public void setAsyncReturn(AsyncReturn asyncReturn) {
+        this.asyncReturn = asyncReturn;
+    }
+
+    /**
+     * This method decodes access token and compare its expiry time with the current time to decide whether it's
+     * expired.
+     *
+     * @param accessToken Access token which needs to be evaluated
+     * @return A boolean value indicating whether the token is expired
+     * @throws ParseException {@link ParseException}
+     */
+    private boolean isTokenExpired(String accessToken) throws ParseException {
+
+        SignedJWT decodedToken = SignedJWT.parse(accessToken);
+        Date expiryDate = (Date) decodedToken.getJWTClaimsSet().getClaim(JWT_EXP_CLAIM);
+        LocalDateTime expiryTimestamp = LocalDateTime.ofInstant(expiryDate.toInstant(), ZoneId.systemDefault());
+        return LocalDateTime.now().isAfter(expiryTimestamp);
+    }
+
+    public void setConsumerKey(String consumerKey) throws SecretManagementException {
+        this.consumerKey = getResolvedSecret(consumerKey);
+    }
+
+    public void setConsumerSecret(String consumerSecret) throws SecretManagementException {
+        this.consumerSecret = getResolvedSecret(consumerSecret);
+    }
+
+    public void setTokenEndpoint(String tokenEndpoint) {
+        this.tokenEndpoint = tokenEndpoint;
+    }
+
+    public void setScopes(String scopes) {
+        this.scopes = scopes;
+    }
+
+    public String getTokenEndpoint() {
+        return tokenEndpoint;
+    }
+
+    public String getConsumerKey() {
+        return consumerKey;
+    }
+
+    public String getConsumerSecret() {
+        return consumerSecret;
+    }
+
+    public String getScopes() {
+        return scopes;
+    }
+
+    @Override
+    public HttpUriRequest applyAuth(HttpUriRequest request, AuthConfigModel authConfigModel)
+            throws FrameworkException, SecretManagementException {
+
+        this.apiAccessTokenCache = APIAccessTokenCache.getInstance();
+        Map<String, Object> properties = authConfigModel.getProperties();
+        validateRequiredProperties(properties);
+
+        setConsumerKey(properties.get(CONSUMER_KEY_VARIABLE_NAME).toString());
+        setConsumerSecret(properties.get(CONSUMER_SECRET_VARIABLE_NAME).toString());
+        setTokenEndpoint(properties.get(TOKEN_ENDPOINT).toString());
+        setScopes(properties.containsKey(SCOPES) ? properties.get(SCOPES).toString() : null);
+
+        String accessToken = getAccessToken();
```

Comment:
Does access token call adheres to the same timeouts and retry attempt count ?
or what's the strategy

Reply:
Same timeouts as in the external call

but the retry counts will be 2 
which is not configurable at the moment.

https://github.com/wso2-extensions/identity-conditional-auth-functions/blob/c29ba564cad40717216931e04224f3e60b3d431f/components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/util/ClientCredentialAuthConfig.java#L81

Reply:
Rather than having a hardcoded value let's adhere to the same retry count. Because when it comes to request time out etc, we are adhering to the same configs